File: tests/refs/linear_search.cpp
// Reference C++ driver for tests/cases/linear_search.json.
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

int linearSearch(const std::vector<long long>& arr, long long target) {
    for (std::size_t i = 0; i < arr.size(); ++i)
        if (arr[i] == target) return static_cast<int>(i);
    return -1;
}

int main() {
    std::string line, caseName;
    std::vector<long long> arr;
    long long target = 0, expected = 0;
    bool haveInput = false, haveExpected = false;
    int passed = 0, failed = 0;

    auto run = [&]() {
        if (caseName.empty() || !haveInput || !haveExpected) return;
        long long got = linearSearch(arr, target);
        if (got == expected) { std::cout << "PASS linear_search :: " << caseName << "\n"; ++passed; }
        else { std::cout << "FAIL linear_search :: " << caseName << "\n  expected: " << expected << "\n  got: " << got << "\n"; ++failed; }
        caseName.clear(); arr.clear(); haveInput = haveExpected = false;
    };

    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        std::istringstream iss(line);
        std::string tag; iss >> tag;
        if (tag == "CASE") iss >> caseName;
        else if (tag == "INPUT") {
            arr.clear();
            std::size_t n; iss >> n;
            for (std::size_t i = 0; i < n; ++i) { long long v; iss >> v; arr.push_back(v); }
            iss >> target;
            haveInput = true;
        } else if (tag == "INT") { iss >> expected; haveExpected = true; }
    }
    run();
    std::cout << "TOTAL: " << passed << " passed, " << failed << " failed\n";
    return failed == 0 ? 0 : 1;
}

File: tests/refs/linear_search.rs
// Reference Rust driver for tests/cases/linear_search.json.
use std::io::{self, BufRead};

fn linear_search(arr: &[i64], target: i64) -> i64 {
    for (i, &x) in arr.iter().enumerate() {
        if x == target { return i as i64; }
    }
    -1
}

fn main() {
    let stdin = io::stdin();
    let mut case_name = String::new();
    let mut arr: Vec<i64> = Vec::new();
    let mut target: i64 = 0;
    let mut expected: i64 = 0;
    let mut hi = false; let mut he = false;
    let mut p = 0u32; let mut f = 0u32;

    let mut flush = |name: &mut String, arr: &mut Vec<i64>, t: i64, e: i64,
                     hi: &mut bool, he: &mut bool, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *hi && *he {
            let got = linear_search(arr, t);
            if got == e { println!("PASS linear_search :: {}", name); *p += 1; }
            else { println!("FAIL linear_search :: {}\n  expected: {}\n  got: {}", name, e, got); *f += 1; }
        }
        name.clear(); arr.clear(); *hi = false; *he = false;
    };

    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut case_name, &mut arr, target, expected, &mut hi, &mut he, &mut p, &mut f); continue; }
        let mut parts = line.split_whitespace();
        let tag = parts.next().unwrap_or("");
        match tag {
            "CASE" => { case_name = parts.next().unwrap_or("").to_string(); }
            "INPUT" => {
                arr.clear();
                let n: usize = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                for _ in 0..n { if let Some(t) = parts.next() { if let Ok(v) = t.parse::<i64>() { arr.push(v); } } }
                target = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                hi = true;
            }
            "INT" => { expected = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0); he = true; }
            _ => {}
        }
    }
    flush(&mut case_name, &mut arr, target, expected, &mut hi, &mut he, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
